vip-helpers/class-vip-media-library-caching.php

```diff
@@ -0,0 +1,210 @@
+<?php
+/**
+ * Media library caching helpers.
+ * 
+ * @package vip-helpers
+ */
+
+ // phpcs:disable WordPress.DB.DirectDatabaseQuery.DirectQuery
+
+namespace Automattic\VIP\Helpers;
+
+/**
+ * VIP Media Library Caching class.
+ */
+class VIP_Media_Library_Caching {
+	public const MINIMUM_WORDPRESS_VERSION          = '6.4';
+	public const AVAILABLE_MIME_TYPES_CACHE_KEY     = 'vip_available_mime_types_';
+	public const USING_DEFAULT_MIME_TYPES_CACHE_KEY = 'vip_using_default_mime_types_';
+	public const MAX_POSTS_TO_QUERY                 = 100000;
+
+	/**
+	 * Class constructor.
+	 */
+	public function __construct() {
+		global $wp_version;
+
+		if ( ! defined( 'VIP_DISABLE_MIME_TYPE_CACHING' ) &&
+			isset( $wp_version ) &&
+			version_compare( $wp_version, self::MINIMUM_WORDPRESS_VERSION, '>=' ) ) {
+			$this->enable_post_mime_types_caching();
+		}
+	}
+
+	/**
+	 * Enable MIME type caching.
+	 */
+	private function enable_post_mime_types_caching() {
+		add_filter( 'pre_get_available_post_mime_types', array( $this, 'get_cached_post_mime_types' ), 10, 2 );
+		add_action( 'add_attachment', array( $this, 'update_post_mime_types_cache_on_add', 10, 1 ) );
+		add_action( 'attachment_updated', array( $this, 'update_post_mime_types_cache_on_edit' ), 10, 3 );
+		add_action( 'delete_attachment', array( $this, 'update_post_mime_types_cache_on_delete' ), 10, 3 );
+	}
+
+	/**
+	 * Get cached results for get_available_post_mime_types() to avoid a query on every page load.
+	 * 
+	 * @param string[]|null $filtered_mime_types An array of MIME types. Default null.
+	 * @param string        $type                The post type name. Usually 'attachment' but can be any post type.
+	 * @return array An array of MIME types.
+	 */
+	public function get_cached_post_mime_types( $filtered_mime_types, $type ) {
+		global $wpdb;
+
+		$cache_key  = self::AVAILABLE_MIME_TYPES_CACHE_KEY . $type;
+		$mime_types = wp_cache_get( $cache_key );
+
+		if ( false === $mime_types ) {
+			$attachment_count = $wpdb->get_var( $wpdb->prepare( "SELECT COUNT(1) FROM {$wpdb->posts} WHERE post_type = %s", $type ) );
+			$use_defaults     = $attachment_count > self::MAX_POSTS_TO_QUERY;
+
+			if ( $use_defaults ) {
```

Comment:
What are your thoughts on caching this value with an expiry time to prevent it from running a query every time if we don't have anything returned in `$mime_types`?

Reply:
`$use_defaults` is being cached below, so the other filters are aware, but here, it's gated inside the `$mime_types` cache generation in case we have a change in the number of attachments that crosses the `MAX_POSTS_TO_QUERY` threshold, which should effectively change the expected behavior. But maybe I'm overthinking it for some edge cases.